phpstan.el

```diff
@@ -28,19 +28,85 @@
 ;; https://github.com/phpstan/phpstan
 
 ;;; Code:
+(require 'php-project)
 (require 'flycheck nil)
 
+
+;; Variables:
+
+;;;###autoload
+(progn
+  (defvar phpstan-config-file nil)
+  (make-variable-buffer-local 'phpstan-config-file)
+  (put 'phpstan-config-file 'safe-local-variable
+       #'(lambda (v) (if (consp v)
+                         (and (eq 'root (car v)) (stringp (cdr v)))
+                       (null v) (stringp v)))))
+
+;;;###autoload
+(progn
+  (defvar phpstan-level "0")
+  (make-variable-buffer-local 'phpstan-level)
+  (put 'phpstan-level 'safe-local-variable
+       #'(lambda (v) (or (null v)
+                         (integerp v)
+                         (and (stringp v)
+                              (string-match-p "\\`[1-9][0-9]*\\'" v))))))
+
+;;;###autoload
+(progn
+  (defvar phpstan-executable nil)
+  (make-variable-buffer-local 'phpstan-executable)
+  (put 'phpstan-executable 'safe-local-variable
+       #'(lambda (v) (if (consp v)
+                         (and (eq 'root (car v)) (stringp (cdr v)))
+                       (null v) (stringp v)))))
+
+;; Functions:
+(defun phpstan-get-config-file ()
+  "Return path to phpstan configure file or `NIL'."
+  (if phpstan-config-file
+      (if (and (consp phpstan-config-file)
+               (eq 'root (car phpstan-config-file)))
+          (expand-file-name (cdr phpstan-config-file) (php-project-get-root-dir))
+        phpstan-config-file)
+    (cl-loop for name in '("phpstan.neon" "phpstan.neon.dist")
+             for file = nil
+             for dir  = (locate-dominating-file default-directory name)
+             if dir
+             return (expand-file-name name dir))))
+
+(defun phpstan-get-level ()
+  "Return path to phpstan configure file or `NIL'."
+  (cond
+   ((null phpstan-level) "0")
+   ((integerp phpstan-level) (int-to-string phpstan-level))
+   (t phpstan-level)))
+
+(defun phpstan-get-executable ()
+  "Return PHPStan excutable file."
+  (let ((executable (or phpstan-executable '(root . "vendor/bin/phpstan"))))
+    (when (and (consp executable)
+               (eq 'root (car executable)))
+      (setq executable
+            (expand-file-name (cdr executable) (php-project-get-root-dir))))
+    (if (file-exists-p executable)
+        executable
+      (if (executable-find "phpstan")
+          "phpstan"
+        (error "PHPStan executable not found")))))
+
 ;;;###autoload
 (when (featurep 'flycheck)
-  (flycheck-define-checker phpstan-checker
+  (flycheck-define-checker phpstan
     "PHP static analyzer based on PHPStan."
-    :command ("phpstan" 
-              "analyze" 
-              "--no-progress"
-              "--errorFormat=raw" 
+    :command ("php" (eval (phpstan-get-executable))
```

Comment:
Since `phpstan` is generally a PHP script (or Phar archive), this command works fine.  Therefore, it is not possible to specify shell scripts that are not PHP or "real" executables.

Reply:
Is this a feature though?  I run phpstan through docker so this would not work for me for example.  And there are many people who do so as well: https://github.com/phpstan/docker-image

Reply:
Thanks.  This problem was solved with the hack of `phpstan-enabled-and-set-flycheck-variable`.  It is an evil side effect.